Integration tests

unglaublicherdude opened the pull request:
Hi,

I have created a mechanic to run integration/smoke tests against cowrie. 

**The Idea**: Thanks to docker you can easily start a container with the fully functional cowrie services within and run tests of any kind against the running container and with a running cowrie service. So just write unit-tests that do ssh into the service and run commands against it.

With this pull-request I simply submit a way to build the cowrie docker container (placed in the root directory of the repository), start the container and run tests against the container.

**The implemented tests do test**:

Implicit
- Does the docker container build?

Explicit
- Does the port open?
- Does ssh work?
- Do the commands work?`

**Important**: All the tests are not an assertion of an output when a specific input is given because that would be unit tests. These tests do simply test if something breaks when running commands against cowrie.

You can run the tests in a virtualenv directly on your machine and with debugging in pycharm, but when not debugging I would really recommend you to run the tests in a docker-container.

Just build it
```
docker build -t cowrie/test -f IntegrationTests.Dockerfile .
```

And run it
```
docker run \
    -v $(pwd):/app \
    -v /var/run/docker.sock:/var/run/docker.sock \
    cowrie/test \
        pytest -v IntegrationTests
```

**Important:**
There are actually 4 tests that are "broken". Theses tests are broken, because the cowrie service does actually break with an exception, when running base64 or date without any parameter.

**Future Idea**: When this pull-request will be accepted I plan to build more kind of tests

- Smoke Tests with tty logs - I plan to create a tests, that can replay tty-logs against the docker container. My thought here was, that if you get an error-report where you can reproduce the error with a specific sequence of command, you can just put it into this tests and solve the problem. The test will get  green and future tests will always check that the problem does not occur again. 
- Output tests - for testing the different cowrie outputs
- Tests for web downloads - do the downloaded urls actually land in the configured directory?

I did not integrate it into the travis-ci because I think you should first try the tests out. Also @craneworks offered me to do the travis-struff, when the pull request gets merged.

Cheers,
Matthias

unglaublicherdude committed:
Simple Integration-Test. Testing  if the port is open after the container is started.

unglaublicherdude committed:
Simple Integration-Test. Testing  if the port is open after the container is started.

unglaublicherdude committed:
Removed the pycache files.

unglaublicherdude committed:
Removed the pycache files.

unglaublicherdude committed:
Implemented ssh-connection tests.

With this commit I submited:
* A quick fix to make the docker-container actually start
* 2 Tests against the ssh-port of the running docker container.
* I started a requirements.txt just for the integration tests, as we need packages here that are not needed for the other tests. keep it seperate for seperate docker images.

unglaublicherdude committed:
Implemented ssh-connection tests.

With this commit I submited:
* A quick fix to make the docker-container actually start
* 2 Tests against the ssh-port of the running docker container.
* I started a requirements.txt just for the integration tests, as we need packages here that are not needed for the other tests. keep it seperate for seperate docker images.

unglaublicherdude committed:
Just some flake8 fixes.

unglaublicherdude committed:
Just some flake8 fixes.

unglaublicherdude committed:
Seperated by test-type.

unglaublicherdude committed:
Seperated by test-type.

unglaublicherdude committed:
Flake-fixes. Or how I call them ->flixes

unglaublicherdude committed:
Flake-fixes. Or how I call them ->flixes

unglaublicherdude committed:
Make Modules out of it.

I placed some init-files because I refactored the starting of the containers. Meaing I will do the start in a central class with a static method. To import this class I put everything into modules.

unglaublicherdude committed:
Make Modules out of it.

I placed some init-files because I refactored the starting of the containers. Meaing I will do the start in a central class with a static method. To import this class I put everything into modules.

unglaublicherdude committed:
Added a short readme.
I also have to put more requirements into it.

unglaublicherdude committed:
Added a short readme.
I also have to put more requirements into it.

unglaublicherdude committed:
working docker tests

unglaublicherdude committed:
working docker tests

unglaublicherdude committed:
fixed typo

unglaublicherdude committed:
fixed typo

unglaublicherdude committed:
Merge branch 'master' into IntegrationTests

unglaublicherdude committed:
fixed flake-error: unused import DateTime

unglaublicherdude committed:
fixed flake-error: unused import DateTime

unglaublicherdude committed:
Some more Linting

unglaublicherdude committed:
Some more Linting

unglaublicherdude committed:
Maintainer is deprecated.

unglaublicherdude committed:
Merged Dockerfile.

unglaublicherdude committed:
Merge from Master

unglaublicherdude committed:
Working version after merge, from upstream.
* Do check the log-output also on the ssh-test
* Provide a more detailed readme
* Flake Import Ordering (actually the dockerfile will not build, when the import order is wrong)

craneworks commented on IntegrationTests.Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3-alpine3.8
```
I think it's better to stay currently on the same python version as in our release image which is python2

unglaublicherdude replied:
That does not really matter to be honest. This tests are not working with any of the code of the actual project and as it is new code I would really stay with python3 here. 

Python2 is deprecated since forever.

craneworks replied:
Ah, true!

craneworks commented on src/cowrie/commands/yum.py:
```diff
@@ -1,9 +1,3 @@
-# Copyright (c) 2009 Upi Tamminen <desaster@gmail.com>
```
Possible that this is a wrong commit? ^^

unglaublicherdude replied:
Oh yes. Even though I really don't like the code - bloating with this useless copyrights.

craneworks replied:
I agree, that's also something on my bucket list.

craneworks commented on IntegrationTests.Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3-alpine3.8
+
+RUN apk add --no-cache gcc build-base libffi-dev openssl-dev
+COPY requirements-integration-tests.txt /tmp
```
The file is so small we could consider putting the items directly into this file.
Less files inside the repository and less layers in the image

unglaublicherdude replied:
That is a really bad idea. If you care for image layers, just squash the image when building on the CI. For local development more layers are always better, because you utilise the caching of the docker daemon.

crane-denny reviewed (changes requested):
I found some small stuff which can/should be changed

unglaublicherdude committed:
re-added the copyright to the yum-file.

unglaublicherdude committed:
Remved useless chown from Dockerfile. Correct rights are already set at the top of the file.

unglaublicherdude committed:
Moved the IntregrationTests files (Dockerfile) into the IntegrationTests
directory.

Remvoed the part of the readme of how to run the tests without docker.
You need docker to run these tests anyway. Why would you not want to run
the testrunner in docker.

unglaublicherdude committed:
Remove requirements installation.

unglaublicherdude commented:
Ah and @craneworks thank you for the input!

micheloosterhof commented:
hi! thanks for submitting the pull request! 

it doesn't build for me because the requirements file referenced in Dockerfile is not there.

Could you rename 'IntegrationTests' to just 'test', please? Readme.md to README.md. 

Your file naming standard uses Camelcase for files, that's not what I use in the rest of Cowrie (it's all lowercase). 

You're using the `pytest` testing framework. For the other tests we are using `unittest` from Twisted (Python). Have you looked at the other test framework to sync these and not have multiple testing frameworks?

Have you considered not accessing docker.sock straight from the container, but using docker-compose to have a client and a server and test that way?

micheloosterhof commented:
Also, I found this: https://docs.travis-ci.com/user/docker
I'm using Travis currently for continuous integration, it can deal with things like multiple docker containers (one client, one server) and the sounds like an excellent use case.

unglaublicherdude commented:
Please don't doch that. It ist way better to have the container orchestrstion in the test code and not in an additional tool Like docker-compose or the travis-ci. That ist the whole reason why libs Like testcontainers exist (really big in Java or C#, somewhat at the beginnjng die Python). You want to have the whole test-process in code.

Also the docker.sock in an container ist pretty standard in the docker-world. I don't know whats wrong with that?

unglaublicherdude commented:
And sorry for not answering but I am actually out of the game by the flu.

mzfr commented:
@micheloosterhof 	Is this PR needed? If yes I can work on this and fix all the remaining errors otherwise you can close this.

mdujava commented on IntegrationTests/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM python:3-alpine3.8
+
+RUN apk add --no-cache gcc build-base libffi-dev openssl-dev
+COPY requirements-integration-tests.txt /tmp
```
```suggestion
COPY requirements-integration-tests.txt /tmp
RUN pip install -r /tmp/requirements-integration-tests.txt
```
